### ISSUE TITLE
Spike: Where do we store outputs from build steps in K8s?

### ISSUE THREAD

**Issue by topherbullock** [concourse/concourse#3740]
From the exploration in concourse/rfcs#22 ; we need to determine a mechanism to store outputs of build steps in order to pass artifacts between each step. 

Here's some questions to find an answer to in this Spike:
- [ ] What mechanism in K8s can we use to get those bits out of the container running a build step?
- [ ] We'll also need to be able to use these as images, so does it make sense to store these in the registry as layers?
   - [ ]  if we use a registry, where should that registry live?
- [ ] Does the proposed solution scale to what Concourse will be running? Do we need GC?

**Comment by chenbh** [comment]
### Exploring using image registries to add arbritary files to any image
- tldr: it is possible to take any arbitary file/directory and add it to any image programatically (no docker required) via api calls and elbow grease
- using gcr (which utilizes gcs as the underlying blobstore)
### Notes
- a "runnable" image is composed of the `manifest`, `config`, and `layers`
- `manifest` has a separate endpoint that can be constructed with the image name and tag
- `config` and `layers` are really just blobs stored somewhere, have to be referenced by digest (sha256)
- `manifest`
  - overview of the image
  - ties together the `config` and the `layers`
  - ```json
    {
      "schemaVersion": 2,
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "config": {
        "mediaType": "application/vnd.docker.container.image.v1+json",
        "size": 1512,
        "digest": "sha256:b2d50e22cf6890c993703735c9cf41855569fcc258bf39ac56d9fab96f787ce0"
      },
      "layers": [
        {
          "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
          "size": 2780373,
          "digest": "sha256:b8141ae1f6634cc1a38d5c04411fbbadccbeb117ee18d1d511fb78aa7cd5c7c7"
        },
        {
          "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
          "size": 160,
          "digest": "sha256:98b6bfc151b4f182c9f9746a77c41fba55660f79b19de95a7a5e7e278e61fb63"
        }
      ]
    }
    ```
- `config`
  - `.rootfs.diff_ids[]` are the sha256 sums of the **uncompressed tarball** for each layer
    - stored oldest first
    - will complain if there are layers without a corresponding diff_id
  - ```json
    {
      "architecture": "amd64",
      "config": {
        ...
      },
      "container": "c07fd716fb4d57f52430af240a3d4e76bf34c5a108772f9789bf2629f3ccb44e",
      "container_config": {
        ...
      },
      "created": "2019-04-09T23:20:00.726573147Z",
      "docker_version": "18.06.1-ce",
      "history": [
        {
          "created": "2019-04-09T23:20:00.566970278Z",
          "created_by": "/bin/sh -c #(nop) ADD file:429aeddcd15b4494dac59a18dd78cc783510d0d9c83e192d6bc32788d451b56e in / "
        },
        {
          "created": "2019-04-09T23:20:00.726573147Z",
          "created_by": "/bin/sh -c #(nop)  CMD [\"/bin/sh\"]",
          "empty_layer": true
        },
        {
          "created": "2019-04-16T14:20:00.726573147Z",
          "created_by": "super janky manual editing yo",
        }
      ],
      "os": "linux",
      "rootfs": {
        "type": "layers",
        "diff_ids": [
          "sha256:12e1a2c4bdb95c62de21bbddcb3f71dadc1d0618315d28f39d06fa9cfc15827d",
          "sha256:ba984ece78babd4d638d43c6180c8e492d8fa74c266e1538345aaa969dc62d99"
        ]
      }
    }
    ```
  - `layers`
    - digest is the sha256 sum of the **compressed tarball**
    - gzip'd tarballs that contain a changeset to the file system
    - any files added here will be added to the rootfs
    - any modified files here will overwrite any previous layers
    - deleted files must be prefixed with `.wh.`
    - https://github.com/moby/moby/blob/master/image/spec/v1.md#creating-an-image-filesystem-changeset
    - ```sh
      $ tar -tzf boo.tgz
      boo/
      boo/foo.txt
      ```
- API
  - `manifest`
    - reference can be either tag or digest
    - GET/PUT/DELETE `/v2/$repo/$image/manifests/$reference`
  - `blob`
    - digest is of the form: `sha256:b2d50e22cf6890c993703735c9cf41855569fcc258bf39ac56d9fab96f787ce0`
    - GET `/v2/$repo/$image/blobs/$digest`
    - POST `/v2/$repo/$image/blobs/uploads/?digest=$digest`
  - deletions
    - gcr does not allow deleting blobs through the API, we would have to go through gcs to do so
    - deleting a manifest via a tag will only untag it
    - deleting a manifest via digest will delete the manifest and cleanup the layers (if they're not in use by other images)

### Demo
https://gist.github.com/chenbh/7cb7b545a535cf6cf04823d250f04efc

**Comment by cirocosta** [comment]
That's great some great work, @chenbh ! Thanks! 

One point regarding the format for the manifest, I'd step away from the one you've pointed out (`v1`) as that's been deprecated AFAIK. 

From what I understand, the manifest v2 (https://docs.docker.com/registry/spec/manifest-v2-2/) is compatible with the OCI spec (https://github.com/opencontainers/image-spec/blob/master/manifest.md) which is the way that container runtimes have been going forward.

Thanks!

**Comment by chenbh** [comment]
@cirocosta there's 2 separate versioned specs that are needed here:

Image Manifest
- identified by `application/vnd.docker.distribution.manifest.v2+json` or `application/vnd.oci.image.manifest.v1+json` (note that the docker is v2 while oci is v1)
- https://github.com/docker/distribution/blob/master/docs/spec/manifest-v2-2.md#image-manifest
- https://github.com/opencontainers/image-spec/blob/master/manifest.md
- Image Manifest Schema Version
  - This is independent of the Manifest version (not confusing at all), either `1` or `2`
  - This is why Docker docs says "Image Manifest V 2, Schema 1" is deprecated and "Image Manifest V 2, Schema 2" is the current one
  - Both of the Manifest versions mentioned above use Schema version 2


Image Config
- identified by `application/vnd.docker.container.image.v1+json` or `application/vnd.oci.image.config.v1+json`
- https://github.com/moby/moby/blob/master/image/spec/v1.2.md#image-json-description
- https://github.com/opencontainers/image-spec/blob/master/config.md

**Comment by cirocosta** [comment]
Oooh I see, thanks for the clarification!

**Comment by ddadlani** [comment]
It may be helpful to look at MezzFS from Netflix as an example of using buckets for build output artifacts. https://medium.com/netflix-techblog/mezzfs-mounting-object-storage-in-netflixs-media-processing-platform-cda01c446ba

**Comment by ClumsyPotato** [comment]
Am i understanding it correct that the requirement 
"We'll also need to be able to use these as images, so does it make sense to store these in the registry as layers"
means you want to be able to store step outputs in images?
If yes, why?

Why not just store them in a persistent volumes?

**Comment by ddadlani** [comment]
@ClumsyPotato for example, Concourse allows image resources fetched in a `get` step to be used as the rootfs for subsequent `task` step containers in the same build. Storing the result of a `get` step in a registry would allow for this behaviour.

**Comment by chenbh** [comment]
After discussing it with @ddadlani, @kcmannem and @nader-ziada, we should be able to use a blob store (either something like gcs or bundle our own) to store input/outputs. 

We might be able to leverage Tekton's [storage resource](https://github.com/tektoncd/pipeline/blob/master/docs/resources.md#storage-resource) to do some of those work for us

**Comment by chenbh** [comment]
# Design decision considerations


## Must have
- store arbitrary binary data
- retrieve arbitrary binary data

## Nice to have
- allow K8s to instantiate a container from a root.tgz file system image

## Criteria
- how easy is it to deploy/manage this subsystem
- how many choices/implementations are offered for this subsystem
- is there a resonable implementation of this subsystem on every major IaaS
- can this be packaged inside Concourse aka hidden away from the user or does the user need to be aware of it ?
- how comfortable are users with storing sensitive info in the subsystem ?


## Assumptions
- Concourse developers use image registry resource to be able to instantiate containers from a root.tgz file, this feature of Concourse isn't leverage by other Concourse users
- Step container isolation at the filesystem level is required

**Comment by chenbh** [comment]
## Exploring using Tekton and using StorageResources to transfer input/outputs
[concourse pipeline](https://gist.github.com/chenbh/f84f5a2be9c57e3dfd3b1374b2fd0c57) (classical booklit)
[tekton pipeline](https://gist.github.com/chenbh/3a428ee05024b3d13c28df22dccc4901)


- used GCS as a blobstore and Tekton's [storage resource](https://github.com/tektoncd/pipeline/blob/master/docs/resources.md#gcs-storage-resource) to transfer files between different (Tekton) Tasks
- implemented (Concourse) resources as a (Tekton) Task using the custom resource type contract
  - more flexibility and the same general concept can be used to map `put` steps and custom resources- found some potential enhancements regarding Tekton resources
- assumed inlined [task configs](https://concourse-ci.org/task-step.html#task-step-config) instead of [task files](https://concourse-ci.org/task-step.html#task-step-file)
  - if we're to use task files, we might have to create a (Tekton) Pipeline to `get` the resource, upload it to GCS, download it to the `web` node, and figure out the task definition from there
- `runAfter` in (Tekton) Pipelines to control sequential/parallel execution of (Tekton) Tasks

## Conclusion
Felt pretty natural to translate the pipeline, no show-stoppers

## Further explorations
- more/custom PipelineResources types
- review contract Concourse provides for container filesystem isolation (Tekton Task Steps share same workspace)
- lifecycle hooks (on_failure, on_error), aka only execute steps if the previous step(s) failed
- stop Tekton from generating a PVC if there's no `from`  condition in any input resources

**Comment by cirocosta** [comment]
Helm seems to be moving towards using OCI Distribution-spec compatible registries as the way of storing & retrieving charts: https://blog.bacongobbler.com/post/2019-01-25-distributing-with-distribution/index.html
